pkg/pgmodel/pgx.go

```diff
@@ -404,76 +435,230 @@ func (p *pgxInserter) getMetricTableName(metric string) (string, error) {
+	select {
+	case req := <-h.input:
+		h.handleReq(req)
+		return true
+	default:
+		return false
+	}
+}
 
-	const batchSize int = 1000
+func (h *InsertHandler) handleReq(req insertDataRequest) bool {
+	needsFlush, pending := h.pending.addReq(req)
+	if needsFlush {
+		h.flushPending(pending)
+		return true
+	}
+	return false
+}
 
-	var needsResponse []insertDataTask
+func (p *PendingBuffer) addReq(req insertDataRequest) bool {
+	p.needsResponse = append(p.needsResponse, insertDataTask{finished: req.finished, errChan: req.errChan})
+	p.batch.sampleInfos = append(p.batch.sampleInfos, req.data...)
+	return len(p.batch.sampleInfos) > flushSize
+}
 
+func (h *InsertHandler) flushTimedOutReqs() {
 	for {
-		req, ok := <-input
-		if !ok {
+		earliest := h.pending.Front()
+		if earliest == nil {
 			return
 		}
-		needsResponse = append(needsResponse, insertDataTask{finished: req.finished, errChan: req.errChan})
-		batch.sampleInfos = append(batch.sampleInfos, req.data...)
 
-	TryReadMore:
-		for len(batch.sampleInfos) < batchSize {
+		earliestPending := earliest.Value.(*PendingBuffer)
+		elapsed := time.Now().Sub(earliestPending.start)
+		if elapsed < flushTimeout {
+			return
+		}
+
+		h.flushPending(earliest)
+	}
+}
+
+func (h *InsertHandler) flushEarliestReq() {
+	earliest := h.pending.Front()
+	if earliest == nil {
+		return
+	}
+
+	h.flushPending(earliest)
+}
+
+func (h *InsertHandler) flushPending(pendingElem *list.Element) {
+	pending := h.pending.Remove(pendingElem)
+
+	_, err := h.conn.CopyFrom(
+		context.Background(),
+		pgx.Identifier{promSchema, pending.metricTable},
+		copyColumns,
+		&pending.batch,
+	)
+
+	for i := 0; i < len(pending.needsResponse); i++ {
+		if err != nil {
 			select {
-			case req := <-input:
-				needsResponse = append(needsResponse, insertDataTask{finished: req.finished, errChan: req.errChan})
-				batch.sampleInfos = append(batch.sampleInfos, req.data...)
+			case pending.needsResponse[i].errChan <- err:
 			default:
-				break TryReadMore
 			}
 		}
-		_, err := p.conn.CopyFrom(
-			context.Background(),
-			pgx.Identifier{promSchema, tableName},
-			copyColumns,
-			&batch,
-		)
-		for i := 0; i < len(batch.sampleInfos); i++ {
-			batch.sampleInfos[i] = nil
-		}
-		batch = SampleInfoIterator{sampleInfos: batch.sampleInfos[:0], sampleIndex: -1, sampleInfoIndex: 0}
-		for i := 0; i < len(needsResponse); i++ {
-			if err != nil {
-				select {
-				case needsResponse[i].errChan <- err:
-				default:
-				}
-			}
-			needsResponse[i].finished.Done()
-			needsResponse[i] = insertDataTask{}
+		pending.needsResponse[i].finished.Done()
+		pending.needsResponse[i] = insertDataTask{}
+	}
+	pending.needsResponse = pending.needsResponse[:0]
+
+	for i := 0; i < len(pending.batch.sampleInfos); i++ {
+		pending.batch.sampleInfos[i] = nil
+	}
+	pending.batch = SampleInfoIterator{sampleInfos: pending.batch.sampleInfos[:0], sampleIndex: -1, sampleInfoIndex: 0}
+
+	h.pending.giveOldBuffer(pending)
```

Comment:
why don't we just call h.pending.Remove here and have that call giveOldBuffer

Reply:
I don't want anything in `pending` that's being inserted.